Rewrite the logic for folding comparisons in PeepholeFoldConstants

Dominator008 opened the pull request:
evaluateComparison() is now based on the three Abstract Operations
defined in the ES6 Specs: Abstract Relational Comparison [1],
Abstract Equality Comparison [2] and Strict Equality Comparison [3].

This fixes a handful of foldable cases that were not folded previously:

Cases like `0 < null` and `'hi' < null` are now folded.

String-string comparisons and string-number comparisions
like `'a' < 'b'` and `1 < '2'` are now folded.

Cases involving `NaN` are now folded.

As a bonus, a test case in testOptimizeSwitch() in
PeepholeRemoveDeadCodeTest now folds.

[1] http://www.ecma-international.org/ecma-262/6.0/#sec-abstract-relational-comparison
[2] http://www.ecma-international.org/ecma-262/6.0/#sec-abstract-equality-comparison
[3] http://www.ecma-international.org/ecma-262/6.0/#sec-strict-equality-comparison

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/google/closure-compiler/1591)
<!-- Reviewable:end -->

Dominator008 commented:
@concavelenz.

blickly commented on src/com/google/javascript/jscomp/PeepholeFoldConstants.java:
```diff
@@ -1075,12 +1075,26 @@ static TernaryValue evaluateComparison(int op, Node left, Node right, boolean us
+          case Token.LE:
+          case Token.GT:
+          case Token.LT:
+            return compareAsNumbers(op, left, right, useTypes);
```
I don't think this is a correct optimization. Consider, for example: `'a'  < 'b'` vs. `+'a' < +'b'`

Dominator008 replied:
Good catch! Need an `else`.

dimvar commented:
LGTM. I'll merge this internally.

dimvar commented:
Undo LGTM. Ben is not sure that this is correct, so I'm reassigning to him and he can take another look and suggest changes.

blickly commented on src/com/google/javascript/jscomp/PeepholeFoldConstants.java:
```diff
@@ -1080,6 +1081,15 @@ static TernaryValue evaluateComparison(int op, Node left, Node right, boolean us
         if (right.isNumber()) {
           return compareAsNumbers(op, left, right, useTypes);
+        } else if (right.isString()) {
+          return compareAsNumbers(op, left, IR.number(NodeUtil.getNumberValue(right)), useTypes);
```
This branch is not correct for SHEQ and SHNE, since, e.g. `'1' !== 1`.
I think that those cases are handled are handled earlier so that this branch isn't reachable for SHEQ and SHNE, but I'd rather that the code be more explicit about what cases are covered and what the ES semantics are. Links into the relevant parts of the spec would also be helpful.

Dominator008 replied:
I added tests for these and made the code explicit. Added a couple of links to the ES6 specs.

blickly commented on src/com/google/javascript/jscomp/PeepholeFoldConstants.java:
```diff
@@ -1064,15 +1064,24 @@ static TernaryValue evaluateComparison(int op, Node left, Node right, boolean us
-              return areStringsEqual(left.getString(), right.getString()).not();
+              return TernaryValue.TRUE;
+          }
+          return compareAsNumbers(op, IR.number(NodeUtil.getNumberValue(left)), right, useTypes);
```
Isn't this whole
`IR.number(NodeUtil.getNumberValue(x))`
extraneous with what compareAsNumbers does anyway?

Dominator008 replied:
Yes. `compareAsNumbers` was being too restrictive about `EQ` and `NE`. I removed the precondition checks that require both operands to be numbers for `EQ` and `NE`.

blickly commented on src/com/google/javascript/jscomp/PeepholeFoldConstants.java:
```diff
@@ -1064,15 +1064,24 @@ static TernaryValue evaluateComparison(int op, Node left, Node right, boolean us
+          return compareStrings(op, left, right);
+        } else if (right.isNumber()) {
+          // See http://www.ecma-international.org/ecma-262/6.0/#sec-strict-equality-comparison
           switch (op) {
```
I'd prefer a switch (and links to the spec) for the cases that can be compared as numbers, rather than the cases that can't.

In fact, can't we merge this case with the following one?

Dominator008 replied:
Done merging. Added a `TODO` saying we could do better, but it requires more changes to the other cases of the left hand side. Will investigate later.

blickly commented on src/com/google/javascript/jscomp/PeepholeFoldConstants.java:
```diff
@@ -1080,6 +1087,22 @@ static TernaryValue evaluateComparison(int op, Node left, Node right, boolean us
+              return compareAsNumbers(op, left, right, useTypes);
+            case Token.EQ:
+            case Token.NE:
+              // TODO(moz): We can fold more cases, but testCommutativeOperators() will fail
```
I'm not sure I follow this TODO. If we did this for non-strings, it would be wrong.
e.g.
0 == null would be folded to true instead of false.

Dominator008 replied:
By "more cases" I meant just booleans (`true` or `false`).

blickly commented on src/com/google/javascript/jscomp/PeepholeFoldConstants.java:
```diff
@@ -1064,15 +1064,22 @@ static TernaryValue evaluateComparison(int op, Node left, Node right, boolean us
-            case Token.SHNE:
             case Token.NE:
-              return areStringsEqual(left.getString(), right.getString()).not();
+              // TODO(moz): We can fold more cases, but testCommutativeOperators() will fail
```
I'm not sure I follow this TODO. If we did this for non-numbers, it would be wrong.
e.g.
'' == null
would be folded to true instead of false.

Dominator008 replied:
By "more cases" I meant just booleans (`true` or `false`).

blickly commented on src/com/google/javascript/jscomp/PeepholeFoldConstants.java:
```diff
@@ -1151,16 +1174,21 @@ private static TernaryValue compareAsNumbers(int op, Node left, Node right, bool
             left.isNumber() && right.isNumber());
         return TernaryValue.forBoolean(lv != rv);
+      // See http://www.ecma-international.org/ecma-262/6.0/#sec-abstract-equality-comparison
+      case Token.EQ:
```
I'm a little bit wary of making this method more permissive. Before, if you called it on `0 == null`, you would get a Preconditions failure, but now you'll get the (incorrect) value of true.

blickly commented on src/com/google/javascript/jscomp/PeepholeFoldConstants.java:
```diff
@@ -1064,15 +1064,22 @@ static TernaryValue evaluateComparison(int op, Node left, Node right, boolean us
+            case Token.LE:
+            case Token.GT:
+            case Token.LT:
+              return compareAsNumbers(op, left, right, useTypes);
```
I dislike this pattern in that comparing a number to (something else) doesn't necessarily mean you should compare them as numbers.

Consider, for example,
`'1' < '0' + '2'`

If we (correctly) improved our implementation of Node.getNumberValue to be able to handle string addition, it would cause this code to become incorrect.

Dominator008 replied:
I agree. I've done a large rewrite that is based on the ES6 specs. Hopefully the code structure is better...

blickly commented:
Overall, we need more tests. A couple of interesting cases:

1) Things with known type, but unknown value, like `+x`, `'' + x`,  or `!x`
2) Things with side effects, like `NaN === foo()`;

Dominator008 commented:
@blickly 
1. I added test cases with known types but unknown values.
2. I added one test with side effects. Note that the first thing `evaluateComparison()` does is bailing out if `left` or `right` may have side effect: https://github.com/Dominator008/closure-compiler/blob/fold/src/com/google/javascript/jscomp/PeepholeFoldConstants.java#L1113-L1116.

blickly commented on src/com/google/javascript/jscomp/PeepholeFoldConstants.java:
```diff
@@ -956,222 +955,185 @@ private Node tryFoldComparison(Node n, Node left, Node right) {
+          Double rv = NodeUtil.getNumberValue(right, useTypes);
+          if (lv != null && rv != null) {
+            return TernaryValue.forBoolean(lv.doubleValue() == rv.doubleValue());
           }
```
Missing a break here, I think

Dominator008 replied:
Done.
